bindinfo/cache.go

```diff
@@ -30,56 +33,165 @@ const (
 	Invalid = "invalid"
 )
 
-// BindMeta stores the basic bind info and bindSql astNode.
-type BindMeta struct {
-	*BindRecord
-	// HintSet stores the set of hints of binding sql.
-	*HintsSet
-}
-
-// cache is a k-v map, key is original sql, value is a slice of BindMeta.
-type cache map[string][]*BindMeta
-
-// BindRecord represents a sql bind record retrieved from the storage.
-type BindRecord struct {
-	OriginalSQL string
-	BindSQL     string
-	Db          string
+// Binding stores the basic bind hint info.
+type Binding struct {
+	BindSQL string
 	// Status represents the status of the binding. It can only be one of the following values:
 	// 1. deleted: BindRecord is deleted, can not be used anymore.
-	// 2. using: BindRecord is in the normal active mode.
+	// 2. using: Binding is in the normal active mode.
 	Status     string
 	CreateTime types.Time
 	UpdateTime types.Time
 	Charset    string
 	Collation  string
+	// hint is the parsed hints, it is used to bind hints to stmt node.
+	Hint *HintsSet
+	// id is the string form of all hints. It is used to uniquely identify different hints.
+	id string
+}
+
+// cache is a k-v map, key is original sql, value is a slice of BindRecord.
+type cache map[string][]*BindRecord
+
+// BindRecord represents a sql bind record retrieved from the storage.
+type BindRecord struct {
+	OriginalSQL string
+	Db          string
+
+	Bindings []Binding
 }
 
-func newBindRecord(row chunk.Row) *BindRecord {
-	return &BindRecord{
-		OriginalSQL: row.GetString(0),
-		BindSQL:     row.GetString(1),
-		Db:          row.GetString(2),
-		Status:      row.GetString(3),
-		CreateTime:  row.GetTime(4),
-		UpdateTime:  row.GetTime(5),
-		Charset:     row.GetString(6),
-		Collation:   row.GetString(7),
+// FirstUsingBinding returns the first binding that has status 'using'.
+// TODO: Remove it when we support using all bindings to adjust the plan.
```

Comment:
What is this TODO mean?

Reply:
We need it because currently optimizer only support adjust plan using 1 binding. In next pr, I will support multiply bindings for optimizer, so it is no longer needed.